### PR TITLE
Bar 34 prepare british imbalanced split

### PR DESCRIPTION
**Pull request by satyarth**

### CODE REVIEW

File: dimred/src/gwas/train_val_split.py [hunk 1]
```diff
@@ -1,224 +1,285 @@
...
+        Args:
+            pca_path (str): Path to PCA eigenvec file computed by plink 2.0
+            cov_pheno_path (str): Path to phenotype and covariates file prepared with ukb_loader
+            pca_cov_path (str): Path where all covariates including PCs will be stored
+            phenotype_path (str): Path with only phenotype data
+
+        """   
+        pca = pandas.read_table(pca_path)
+        cov_pheno = pandas.read_table(cov_pheno_path)
+        cov_columns = list(cov_pheno.columns)[2:-1]
```
Comment (dkolobok) [inline dimred/src/gwas/train_val_split.py:179]: why not have a list of covariates and use column names rather than indices? Indices are not very robust + we might want to consider different sets of covariates

File: dimred/src/gwas/train_val_split.py [hunk 2]
```diff
@@ -1,224 +1,285 @@
...
+        pheno_column = cov_pheno.columns[-1]
+        # print(f'COV_PHENO_COLUMNS are: {cov_columns}')
+        merged = pca.merge(cov_pheno, how='inner', on=['FID', 'IID'])
+        
+        pca_cov = merged.loc[:, ['FID', 'IID'] + [f'PC{i}' for i in range(1, 11)] + cov_columns]
```
Comment (dkolobok) [inline dimred/src/gwas/train_val_split.py:184]: let's have number of PCs as a config parameter

File: dimred/src/gwas/train_val_split.py [hunk 3]
```diff
@@ -1,224 +1,285 @@
...
+            val_path (str): Path to .tsv file with val data. First two columns should be FID, IID
+            columns (List[str]): List of columns to standardize. By default all columns except FID and IID will be standardized. 
+        """    
+        train_data = pandas.read_table(train_path)
+        val_data = pandas.read_table(val_path)
+        test_data = pandas.read_table(test_path)
+
+        scaler = StandardScaler()
+        if columns is None:
+            train_data.iloc[:, 2:] = scaler.fit_transform(train_data.iloc[:, 2:]) # 0,1 are FID, IID
```
Comment (dkolobok) [inline dimred/src/gwas/train_val_split.py:233]: please use column names rather than column indices

File: dimred/src/gwas/train_val_split.py [hunk 4]
```diff
@@ -1,224 +1,285 @@
...
+        """        
+        for fold_index in range(FOLD_COUNT):
+            self.standardize(
+                    self.split.get_pca_cov_path(node_index, fold_index, 'train'),
+                    self.split.get_pca_cov_path(node_index, fold_index, 'val'),
+                    self.split.get_pca_cov_path(node_index, fold_index, 'test'),
+                    covariates
+            )
+
+            # standardize phenotype in phenotype-only file
```
Comment (dkolobok) [inline dimred/src/gwas/train_val_split.py:207]: let's remove this?

**Review by dkolobok — CHANGES_REQUESTED** [review]
Looks great overall! I suggest some minor changes. Couldn't assess some of the files because it's difficult to do without running

File: dimred/src/gwas/train_val_split.py [hunk 5]
```diff
@@ -1,224 +1,285 @@
...
+            columns (List[str]): List of columns to standardize. By default all columns except FID and IID will be standardized. 
+        """    
+        train_data = pandas.read_table(train_path)
+        val_data = pandas.read_table(val_path)
+        test_data = pandas.read_table(test_path)
+
+        scaler = StandardScaler()
+        if columns is None:
+            train_data.iloc[:, 2:] = scaler.fit_transform(train_data.iloc[:, 2:]) # 0,1 are FID, IID
+            val_data.iloc[:, 2:] = scaler.transform(val_data.iloc[:, 2:])
```
Comment (dkolobok) [inline dimred/src/gwas/train_val_split.py:234]: It's generally recommended to normalise test and val datasets using mean and variance from the train dataset instead of calculation mean and variance in-house
https://stackoverflow.com/questions/49444262/normalize-data-before-or-after-split-of-training-and-testing-data

Comment (alex-medvedev-msc) [reply dimred/src/gwas/train_val_split.py:234]: scaler.transform uses mean and variance obtained from train_data using fit_transform method

File: dimred/src/gwas/train_val_split.py (hunk 1 above)
Comment (alex-medvedev-msc) [reply dimred/src/gwas/train_val_split.py:179]: In snakemake splitting version, we will use list of covariates as config parameters

File: dimred/src/gwas/train_val_split.py [hunk 6]
```diff
@@ -1,224 +1,285 @@
...
+        
+        Args:
+            node_index (int): Index of node
+            random_state (int): Fixed random_state for train_test_split sklearn function
+        """    
+        path = self.split.get_source_ids_path(node_index)
+        # we do not need sex here
+        ids = pandas.read_table(path).loc[:, ['FID', 'IID']]
+        train_size, val_size, test_size = int(ids.shape[0]*0.8), int(ids.shape[0]*0.1), int(ids.shape[0]*0.1)
+        train_size += (ids.shape[0] - train_size - test_size - val_size)
```
[the remark this answers is not in the document]
Comment (alex-medvedev-msc) [reply dimred/src/gwas/train_val_split.py:85]: We will lose 1 or 2 samples in this case. You probably meant train_size = int(len(ids) / K) * (K-2)?

File: dimred/src/gwas/train_val_split.py [hunk 7]
```diff
@@ -1,224 +1,285 @@
...
+
+        kfold = KFold(n_splits=10, shuffle=True, random_state=random_state)
```
[the remark this answers is not in the document]
Comment (alex-medvedev-msc) [reply dimred/src/gwas/train_val_split.py:87]: I agree, it will be a config parameter in snakemake version
